plugin/pkg/admission/persistentvolumeclaim/pvcprotection/admission.go

```diff
@@ -0,0 +1,111 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pvcprotection
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/golang/glog"
+
+	admission "k8s.io/apiserver/pkg/admission"
+	"k8s.io/apiserver/pkg/util/feature"
+	api "k8s.io/kubernetes/pkg/apis/core"
+	informers "k8s.io/kubernetes/pkg/client/informers/informers_generated/internalversion"
+	corelisters "k8s.io/kubernetes/pkg/client/listers/core/internalversion"
+	"k8s.io/kubernetes/pkg/features"
+	kubeapiserveradmission "k8s.io/kubernetes/pkg/kubeapiserver/admission"
+	volumeutil "k8s.io/kubernetes/pkg/volume/util"
+)
+
+const (
+	// PluginName is the name of this admission controller plugin
+	PluginName = "PVCProtection"
+)
+
+// Register registers a plugin
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName, func(config io.Reader) (admission.Interface, error) {
+		plugin := newPlugin()
+		return plugin, nil
+	})
+}
+
+// pvcProtectionPlugin holds state for and implements the admission plugin.
+type pvcProtectionPlugin struct {
+	*admission.Handler
+	lister corelisters.PersistentVolumeClaimLister
+}
+
+var _ admission.Interface = &pvcProtectionPlugin{}
+var _ = kubeapiserveradmission.WantsInternalKubeInformerFactory(&pvcProtectionPlugin{})
+
+// newPlugin creates a new admission plugin.
+func newPlugin() *pvcProtectionPlugin {
+	return &pvcProtectionPlugin{
+		Handler: admission.NewHandler(admission.Create),
+	}
+}
+
+func (c *pvcProtectionPlugin) SetInternalKubeInformerFactory(f informers.SharedInformerFactory) {
+	informer := f.Core().InternalVersion().PersistentVolumeClaims()
+	c.lister = informer.Lister()
+	c.SetReadyFunc(informer.Informer().HasSynced)
+}
+
+// ValidateInitialization ensures lister is set.
+func (c *pvcProtectionPlugin) ValidateInitialization() error {
+	if c.lister == nil {
+		return fmt.Errorf("missing lister")
+	}
+	return nil
+}
+
+// Admit sets finalizer on all PVCs. The finalizer is removed by
```

Comment:
Add two comments

1. the full feature requires all kubelets to be running at least the version this was introduced + have the feature gate enabled (and what the consequences are if the kubelets aren't enabled)
2. the change to user behavior - PVC deletion used to be instant, now it can be delayed indefinitely, which may break user scripting and cluster workflow automation.

I assume all of the dynamic provisioner plugins will be updated to be tolerate a delete call not actually deleting the underlying resource?  This was the source of many errors on other resources that we changed like this.

Reply:
> I assume all of the dynamic provisioner plugins will be updated to be tolerate a delete call not actually deleting the underlying resource?

As it is already implemented now, storage assets are removed only after corresponding PVCs are deleted. If we postpone deletion of a PVC in this PR, we automatically postpone deletion of the storage asset, no change needed.

Reply:
> Add two comments

Do you mean comment in the admission plugin? That feels out of place. Adding to kubelet check.

> 1. what the consequences are if the kubelets aren't enabled

The same as now, kubelet may run a pod whose PVC was marked as deleted or deleted. There is a tiny window of opportunity when scheduler schedules a pod, user deleted a PVC and the controller removed finalizer on PVC, so it should not happen that often.

Added a comment into kubelet

> 2. the change to user behavior

Yes, it is.